sdk/core/core/inc/az_str.h

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_STR_H
+#define AZ_STR_H
+
+#include <az_assert.h>
+
+#include <az_static_assert.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+#include <limits.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// An immutable string slice.
+typedef struct {
+  // Points to the first byte (character).
+  uint8_t const *begin;
+  // A numer of octets (bytes) in the string.
+  size_t size;
+} az_const_str;
+
+// A muttable string slice.
+typedef struct {
+  uint8_t *begin;
+  size_t size;
+} az_str;
+
+AZ_STATIC_ASSERT(CHAR_BIT == 8);
+
+// A size of the string literal.
+// Details: to make sure that `S` is a `string literal`, we are appending `""` to `S`.
+#define AZ_STRING_LITERAL_LEN(S) (sizeof(S "") - 1)
+
+#define AZ_CONST_STR_DECL(NAME, STRING_LITERAL) \
+  static az_const_str const NAME = { .begin = STRING_LITERAL, .size = AZ_STRING_LITERAL_LEN(STRING_LITERAL) };
+
+#define AZ_CONST_STR(STRING_LITERAL) \
+  (az_const_str){ .begin = STRING_LITERAL, .size = AZ_STRING_LITERAL_LEN(STRING_LITERAL) }
+
+inline uint8_t const *az_const_str_ptr(az_const_str const buffer, size_t const index) {
+  AZ_ASSERT(index <= buffer.size);
+  return buffer.begin + index;
+}
+
+inline uint8_t az_const_str_item(az_const_str const buffer, size_t const index) {
+  AZ_ASSERT(index < buffer.size);
+  return buffer.begin[index];
+}
+
+inline uint8_t const *az_const_str_end(az_const_str const buffer) {
+  return az_const_str_ptr(buffer, buffer.size);
+}
+
+inline size_t az_const_str_index(az_const_str const buffer, uint8_t const *const p) {
```

Comment:
I don't think we should need this either. The caller can keep track of how far into a slice they are as they do processing. One of the main reason for introducing slices is to NOT use pointers and to avoid pointer arithmetic. The caller should try to avoid having a 'p' and, if they do, then they can't pass it to you to call this function.